File: huntober/reverse-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */

// iterative, runtime beats 93.87%, memory beats 26.17%
var reverseList = function(head) {
  // initialize reversed as null
  // loop over the input list
    // reassign reversed to be a new node with current head val and reversed as next
    // reassign head as head next
  // return reversed
  let reversed = null;
  while (head) {
    reversed = new ListNode(head.val, reversed);
    head = head.next;
  }
  return reversed;
};

// iterative, runtime beats 74.20%, memory beats 86.73%
var reverseList = function(head) {
  // initialize a prev var as null
  // create a loop to run as long as head is defined
      // initialize a temp var to hold head next
      // reassign head next to prev value
      // reassign prev to head
      // reassign head to temp
  // return prev
  let prev = null;
  while (head) {
      const temp = head.next;
      [head.next, prev, head] = [prev, head, temp];
  }
  return prev;
};

// recursive, runtime beats 96.53%, memory beats 5.28%
var reverseList = function(head, prev = null) {
  // base case
    // if head is null
      // return prev
  // recursive case
    // establish new current node to be head next
    // set head next to prev
    // return recursive call to reverseList
      // where 1st arg is new current node (old head next)
      // and 2nd arg is original head
  if (!head) return prev;
  return reverseList(head.next, (head.next = prev, head));
};

console.log(reverseList([1,2,3,4,5]), [5,4,3,2,1]);
console.log(reverseList([1,2]), [2,1]);
console.log(reverseList([]), []);

File: leetcode/java/easy/ReverseLinkedList.java
package leetcode.java.easy;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 *
 * Parameters: list node
 * Returns: list node
 * Examples:
 *    Input: head = [1,2,3,4,5]
 *    Output: [5,4,3,2,1]
 *
 *    Input: head = [1,2]
 *    Output: [2,1]
 *
 *    Input: head = []
 *    Output: []
 * Pseudocode:
 *    iterative
 *      prev var initialized to null
 *      loop while head exists
 *        next var initialized to current head next
 *        reassign current head's next to prev
 *        reassign prev to current head
 *        reassign head to next
 *      return prev
 *
 *    recursive
 *      define method that takes in single ListNode param (head)
 *        return call to version of method that takes in 2 params
 *          passing in head as head and null as prev
 *      define method that takes in 2 ListNode params (head & prev)
 *        base case
 *          if head is null
 *            return prev
 *        recursive case
 *          initialize next var to store original head next
 *          reassign head next to prev
 *          return recursive call with next as head param and head as prev param
 */

class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

// iterative
// runtime 0 ms beats 100%, memory 41.3 mb beats 74.89%
class ReverseLinkedListIterativeSolution {
  public ListNode reverseList(ListNode head) {
      ListNode prev = null;
      while (head != null) {
          ListNode next = head.next;
          head.next = prev;
          prev = head;
          head = next;
      }
      return prev;
  }
}

// recursive
// runtime 0 ms beats 100%, memory 41.9 mb beats 21.86%
class ReverseLinkedListRecursiveSolution {
    public ListNode reverseList(ListNode head) {
        return reverseList(head, null);
    }
    public ListNode reverseList(ListNode head, ListNode prev) {
        if (head == null) return prev;
        ListNode next = head.next;
        head.next = prev;
        return reverseList(next, head);
    }
}